src/program/lwaftr/doc/README.configuration.md
# lwAFTR configuration

The lwAFTR is configured by a text file.  Here's an example:

```
aftr_ipv4_ip = 10.10.10.10
aftr_ipv6_ip = 8:9:a:b:c:d:e:f
aftr_mac_b4_side = 22:22:22:22:22:22
aftr_mac_inet_side = 12:12:12:12:12:12
next_hop6_mac = 44:44:44:44:44:44
binding_table = path/to/binding-table.txt
hairpinning = true
icmpv6_rate_limiter_n_packets=3e5
icmpv6_rate_limiter_n_seconds=4
inet_mac = 68:68:68:68:68:68
ipv4_mtu = 1460
ipv6_mtu = 1500
policy_icmpv4_incoming = ALLOW
policy_icmpv6_incoming = ALLOW
policy_icmpv4_outgoing = ALLOW
policy_icmpv6_outgoing = ALLOW
v4_vlan_tag = 1234
v6_vlan_tag = 42
vlan_tagging = true
ipv4_ingress_filter = "ip src 192.168.0.1"
ipv4_egress_filter = "ip dst 8.8.8.8"
ipv6_ingress_filter = "ip6 src fe80::1"
ipv6_egress_filter = "ip dst 2001:4860:4860::8888"
```

The lwAFTR is associated with two physical network cards. One of these cards
faces the internet; traffic over it is IPv4. The other faces the IPv6-only
internal network, and communicates primarily with B4s.

## Line-by-line explanation

### L2 and L3 addresses of the lwAFTR

First, we set the IP and MAC addresses for both interfaces:

```
aftr_ipv4_ip = 10.10.10.10
aftr_ipv6_ip = 8:9:a:b:c:d:e:f
aftr_mac_b4_side = 22:22:22:22:22:22
aftr_mac_inet_side = 12:12:12:12:12:12
```

This associates **12:12:12:12:12:12** and **10.10.10.10** with the
internet-facing NIC, and **8:9:a:b:c:d:e:f** and **22:22:22:22:22:22** with the
NIC facing the internal network.

### L2 next hops

Normally you might expect to just set a default IPv4 and IPv6 gateway
and have the lwAFTR figure out the next hop ethernet addresses on its
own.  However the lwAFTR doesn't support
[ARP](https://en.wikipedia.org/wiki/Address_Resolution_Protocol) yet.

The lwAFTR assumes that it will talk directly to only one host on each
side, and provides these configuration options to specify the L2
addresses of those hosts.

```
next_hop6_mac = 44:44:44:44:44:44
inet_mac = 68:68:68:68:68:68
```

The lwAFTR can talk to any host, but assumes that the above ones are the
next hop.

The lwAFTR does support
[NDP](https://en.wikipedia.org/wiki/Neighbor_Discovery_Protocol),
though, so the `next_hop6_mac` entry is optional.  See [the NDP
documentation](./README.ndp.md), for more.

### The binding table

```
binding_table = path/to/binding-table.txt
```

See [README.bindingtable.md](README.bindingtable.md) for binding table
details.  Note that you can compile the binding table beforehand; again,
see [README.bindingtable.md](README.bindingtable.md).

### Hairpinning

```
hairpinning = true
```

Configurable hairpinning is a requirement of [RFC
7596](https://tools.ietf.org/html/rfc7596); it can be true or false.

### Rate-limiting of ICMP error messages

```
icmpv6_rate_limiter_n_packets=3e5
icmpv6_rate_limiter_n_seconds=4
```

ICMPv6 rate limiting is mandated by several RFCs. This example says that the
lwAFTR can send at most 300,000 (3 * 10^5) ICMPv6 packets per 4 seconds.
Lower values are recommended for non-experimental use.

### MTU

```
ipv4_mtu = 1460
ipv6_mtu = 1500
```

The MTU settings are used to determine whether a packet needs to be
fragmented.  The current MTU handling is otherwise underdeveloped.  It
is not dynamically updated on receiving ICMP packet too big messages.

### ICMP handling policies

```
policy_icmpv4_incoming = ALLOW
policy_icmpv6_incoming = ALLOW
policy_icmpv4_outgoing = ALLOW
policy_icmpv6_outgoing = ALLOW
```

The lwAFTR can be configured to `ALLOW` or `DROP` incoming and outgoing
ICMPv4 and ICMPv6 messages. If a finer granularity of control is
desired, contact the development team via github or email.

### VLAN tagging

```
v4_vlan_tag = 1234
v6_vlan_tag = 42
vlan_tagging = tru
```

Enable/disable 802.1Q Ethernet tagging with 'vlan_tagging'.

If it is enabled, set one tag per interface to tag outgoing packets with, and
assume that incoming packets are tagged. If it is 'false', v4_vlan_tag and
v6_vlan_tag are currently optional (and unused).

Values of `v4_vlan_tag` and `v6_vlan_tag` represent the identifier value in a
VLAN tag. It must be a value between 0 and 4095.

More sophisticated support, including for mixes of tagged/untagged packets,
will be provided upon request.

### Ingress and egress filters

```
ipv4_ingress_filter = "ip src 192.168.0.1"
ipv4_egress_filter = "ip dst 8.8.8.8"
ipv6_ingress_filter = "ip6 src fe80::1"
ipv6_egress_filter = "ip dst 2001:4860:4860::8888"
```

In the example configuration these entries are commented out by the `#`
character.  If uncommented, the right-hand-side should be a
[pflang](https://github.com/Igalia/pflua/blob/master/doc/pflang.md)
filter.  Pflang is the language of `tcpdump`, `libpcap`, and other
tools.  If these options are given, the filters will run on the packets
without vlan tags, if any.

Filter definitions may also be stored in separate config files, one per
filter, and their relative paths go to the right-hand side of these entries.
The relative paths should be prefixed by `<` and are based on the directory
where the main config file is. Example:

```
ipv4_ingress_filter = <ingress4.pf
```


src/apps/lwaftr/conf.lua
module(..., package.seeall)

local ethernet = require("lib.protocol.ethernet")
local ipv4 = require("lib.protocol.ipv4")
local ipv6 = require("lib.protocol.ipv6")
local Parser = require("apps.lwaftr.conf_parser").Parser

policies = {
   DROP = 1,
   ALLOW = 2
}

local function required(key)
   return function(config)
      error('missing required configuration key "'..key..'"')
   end
end

local function required_if(key, otherkey)
   return function(config)
      if config[otherkey] then
         error('missing required configuration key "'..key..'"')
      end
   end
end

local function required_at_least_one_of(key, otherkey)
   return function(config)
      if config[otherkey] == nil then
         error(string.format("At least one of '%s' and '%s' must be specified", key, otherkey))
      end
   end
end

local function default(val)
   return function(config) return val end
end

local lwaftr_conf_spec = {
   parse={
      aftr_ipv4_ip=Parser.parse_ipv4,
      aftr_ipv6_ip=Parser.parse_ipv6,
      aftr_mac_b4_side=Parser.parse_mac,
      aftr_mac_inet_side=Parser.parse_mac,
      next_hop6_mac=Parser.parse_mac,
      binding_table=Parser.parse_file_name,
      hairpinning=Parser.parse_boolean,
      icmpv6_rate_limiter_n_packets=Parser.parse_non_negative_number,
      icmpv6_rate_limiter_n_seconds=Parser.parse_positive_number,
      inet_mac=Parser.parse_mac,
      ipv4_fragment_cache_max=Parser.parse_non_negative_number,
      ipv6_fragment_cache_max=Parser.parse_non_negative_number,
      ipv4_mtu=Parser.parse_mtu,
      ipv6_mtu=Parser.parse_mtu,
      next_hop_ipv6_addr=Parser.parse_ipv6,
      policy_icmpv4_incoming=Parser.enum_parser(policies),
      policy_icmpv4_outgoing=Parser.enum_parser(policies),
      policy_icmpv6_incoming=Parser.enum_parser(policies),
      policy_icmpv6_outgoing=Parser.enum_parser(policies),
      v4_vlan_tag=Parser.parse_vlan_tag,
      v6_vlan_tag=Parser.parse_vlan_tag,
      vlan_tagging=Parser.parse_boolean,
      ipv4_ingress_filter=Parser.parse_string_or_file,
      ipv4_egress_filter=Parser.parse_string_or_file,
      ipv6_ingress_filter=Parser.parse_string_or_file,
      ipv6_egress_filter=Parser.parse_string_or_file,
   },
   defaults={
      aftr_ipv4_ip=required('aftr_ipv4_ip'),
      aftr_ipv6_ip=required('aftr_ipv6_ip'),
      aftr_mac_b4_side=required('aftr_mac_b4_side'),
      aftr_mac_inet_side=required('aftr_mac_inet_side'),
      next_hop6_mac=required_at_least_one_of('next_hop6_mac', 'next_hop_ipv6_addr'),
      binding_table=required('binding_table'),
      hairpinning=default(true),
      icmpv6_rate_limiter_n_packets=default(6e5),
      icmpv6_rate_limiter_n_seconds=default(2),
      inet_mac=required('inet_mac'),
      ipv4_fragment_cache_max=default(10000),  -- ~160MiB of 16KiB packets
      ipv6_fragment_cache_max=default(10000),
      ipv4_mtu=default(1460),
      ipv6_mtu=default(1500),
      next_hop_ipv6_addr = required_at_least_one_of('next_hop_ipv6_addr', 'next_hop6_mac'),
      policy_icmpv4_incoming=default(policies.ALLOW),
      policy_icmpv4_outgoing=default(policies.ALLOW),
      policy_icmpv6_incoming=default(policies.ALLOW),
      policy_icmpv6_outgoing=default(policies.ALLOW),
      v4_vlan_tag=required_if('v4_vlan_tag', 'vlan_tagging'),
      v6_vlan_tag=required_if('v6_vlan_tag', 'vlan_tagging'),
      vlan_tagging=default(false)
   },
   validate=function(parser, config) end
}

function load_lwaftr_config(stream)
   return Parser.new(stream):parse_property_list(lwaftr_conf_spec)
end

function selftest()
   print('selftest: conf')
   local lib = require('core.lib')
   local function string_file(str)
      local pos = 1
      return {
         read = function(self, n)
            assert(n==1)
            local ret
            if pos <= #str then
               ret = str:sub(pos,pos)
               pos = pos + 1
            end
            return ret
         end,
         close = function(self) str = nil end
      }
   end
   local function test(str, expected)
      if not lib.equal(expected, load_lwaftr_config(string_file(str))) then
         error('lwaftr conf parse produced unexpected result; string:\n'..str)
      end
   end
   test([[
            aftr_ipv4_ip = 1.2.3.4
            aftr_ipv6_ip = 8:9:a:b:c:d:e:f
            aftr_mac_b4_side = 22:22:22:22:22:22
            aftr_mac_inet_side = 12:12:12:12:12:12
            next_hop6_mac = 44:44:44:44:44:44
            binding_table = "foo-table.txt"
            hairpinning = false
            icmpv6_rate_limiter_n_packets=6e3
            icmpv6_rate_limiter_n_seconds=2
            inet_mac = 68:68:68:68:68:68
            ipv4_fragment_cache_max = 1004
            ipv6_fragment_cache_max = 1006
            ipv4_mtu = 1460
            ipv6_mtu = 1500
            policy_icmpv4_incoming = ALLOW
            policy_icmpv6_incoming = ALLOW
            policy_icmpv4_outgoing = ALLOW
            policy_icmpv6_outgoing = ALLOW
            v4_vlan_tag = 1092 # 0x444
            v6_vlan_tag = 1638 # 0x666
            vlan_tagging = true
        ]],
      {
         aftr_ipv4_ip = ipv4:pton('1.2.3.4'),
         aftr_ipv6_ip = ipv6:pton('8:9:a:b:c:d:e:f'),
         aftr_mac_b4_side = ethernet:pton("22:22:22:22:22:22"),
         aftr_mac_inet_side = ethernet:pton("12:12:12:12:12:12"),
         next_hop6_mac = ethernet:pton("44:44:44:44:44:44"),
         binding_table = "foo-table.txt",
         hairpinning = false,
         icmpv6_rate_limiter_n_packets=6e3,
         icmpv6_rate_limiter_n_seconds=2,
         inet_mac = ethernet:pton("68:68:68:68:68:68"),
         ipv4_fragment_cache_max = 1004,
         ipv6_fragment_cache_max = 1006,
         ipv4_mtu = 1460,
         ipv6_mtu = 1500,
         policy_icmpv4_incoming = policies['ALLOW'],
         policy_icmpv6_incoming = policies['ALLOW'],
         policy_icmpv4_outgoing = policies['ALLOW'],
         policy_icmpv6_outgoing = policies['ALLOW'],
         v4_vlan_tag = 0x444,
         v6_vlan_tag = 0x666,
         vlan_tagging = true
      }
   )
   local function test_loading_filter_conf_from_file()
      -- Setup the filter conf file.
      local filter_path = os.tmpname()
      local filter_text = 'some pflang filter string'
      assert(lib.writefile(filter_path, filter_text))
      -- Setup the main config file.
      local conf_text = [[
         aftr_ipv4_ip = 1.2.3.4
         aftr_ipv6_ip = 8:9:a:b:c:d:e:f
         aftr_mac_b4_side = 22:22:22:22:22:22
         aftr_mac_inet_side = 12:12:12:12:12:12
         next_hop6_mac = 44:44:44:44:44:44
         binding_table = "foo-table.txt"
         inet_mac = 68:68:68:68:68:68
         ipv4_ingress_filter = <%s
      ]]
      conf_text = conf_text:format(filter_path)
      local conf_table = load_lwaftr_config(string_file(conf_text))
      assert(os.remove(filter_path))
      if conf_table['ipv4_ingress_filter'] ~= filter_text then
         error('lwaftr: filter conf contents do not match; pathname:\n'..filter_path)
      end
   end
   test_loading_filter_conf_from_file()
   print('ok')
end
